csrc/ops/arith.cpp

```diff
@@ -2087,101 +2088,105 @@ TensorView* viewAsScalar(TensorView* inp) {
-}
-
-} // namespace
-
-TensorView* fusedMultiplySum(
-    TensorView* tv_a,
-    TensorView* tv_b,
-    const std::vector<int64_t>& axes,
-    Val* init) {
-  // TODO:
-  //  Validate axis relationships between a and b
-  NVF_CHECK(tv_a->nDims() > 0, "Tried to reduce a 0-dim tensor");
+  NVF_CHECK(
+      axis_mapping.a_axes.size() == axis_mapping.b_axes.size(),
+      "Axis mapping should contain same number of output axes for each operand");
+  const size_t out_dims = axis_mapping.a_axes.size();
+
+  std::unordered_set<size_t> axes_set;
+  for (int64_t axis : axes) {
+    if (axis < 0) {
+      axis += (int64_t)out_dims;
+    }
+    NVF_ERROR(axis >= 0 && axis < (int64_t)out_dims);
+    axes_set.insert((size_t)axis);
+  }
 
   // TODO:
   //  Add tf32 and other mma data types
   //  Add fallback path for non-mma data types.
   NVF_CHECK(
-      tv_a->getDataType().value() == DataType::Half ||
-      tv_a->getDataType().value() == DataType::BFloat16);
-  NVF_CHECK(tv_a->getDataType().value() == tv_b->getDataType().value());
-
-  NVF_CHECK(!axes.empty(), "No reduction axis specified");
+      tv_a->dtype() == DataType::Half || tv_a->dtype() == DataType::BFloat16);
+  NVF_CHECK(tv_a->dtype() == tv_b->dtype());
+  DataType out_dtype = DataType::Float;
 
-  // TODO:
-  //  will lift this in a follow up when we have a
-  //  more generic axes matching.
-  NVF_CHECK(
-      axes.size() == 1, "Single axis reduction only for mma op instantiation.")
+  // Prepare output domain based on domain mapping and IterTypes of inputs
+  std::vector<IterDomain*> out_domain;
+  out_domain.reserve(axis_mapping.a_axes.size());
+  for (size_t i : c10::irange(out_dims)) {
+    int64_t a_pos = axis_mapping.a_axes[i];
+    int64_t b_pos = axis_mapping.b_axes[i];
+    NVF_CHECK(
+        a_pos != -1 || b_pos != -1,
+        "Output axis ",
+        i,
+        " cannot be missing in both operands");
+    NVF_CHECK(
+        a_pos == -1 || (a_pos >= 0 && a_pos < (int64_t)a_logical.size()),
+        "Position ",
+        i,
+        " in output of axis mapping for operand A is ",
+        a_pos,
+        " which is out of bounds for A which has dimension ",
+        a_logical.size());
+    NVF_CHECK(
+        b_pos == -1 || (b_pos >= 0 && b_pos < (int64_t)b_logical.size()),
+        "Position ",
+        i,
+        " in output of axis mapping for operand B is ",
+        b_pos,
+        " which is out of bounds for B which has dimension ",
+        b_logical.size());
+    IterDomain* a_id = a_pos == -1 ? nullptr : a_logical[(size_t)a_pos];
+    IterDomain* b_id = b_pos == -1 ? nullptr : b_logical[(size_t)b_pos];
+
+    bool a_concrete = a_id == nullptr ? false : !a_id->isBroadcast();
+    bool b_concrete = b_id == nullptr ? false : !b_id->isBroadcast();
+    // NOTE: we can have !a_concrete && !b_concrete if there are broadcast batch
+    // dims
+
+    // Check for K dimensions
+    bool is_reduction = a_concrete && b_concrete && axes_set.count(i);
```

Comment:
Is it possible to have `a_concrete && !b_concrete && axes_set.count(i)`? Should we error out on this case?

Reply:
Great point. I will validate that any K dimensions are concrete in both inputs.